[RFC]vim-patch:8.0.{1123,1125,1138,1139,1142,1292,1334,1375},8.1.1264

janlazo opened the pull request:
Decoupled from https://github.com/neovim/neovim/pull/11997 to focus on menu,toolbar patches.

Patch 8.0.1375 fails the functional tests. I added the w_winbar_height code for Vim patching but it's dead code for Neovim.

janlazo committed:
vim-patch:8.0.1123: cannot define a toolbar for a window

Problem:    Cannot define a toolbar for a window.
Solution:   Add a window-local toolbar.
https://github.com/vim/vim/commit/1b9645de3c05f37b5c30e78f999351b0cf486ade

"WinBar" code in "screen.c" was not ported.

Fix https://github.com/neovim/neovim/issues/11513#issuecomment-562012827

janlazo committed:
vim-patch:8.0.1125: wrong window height when splitting window with window toolbar

Problem:    Wrong window height when splitting window with window toolbar.
Solution:   Add or subtract the window toolbar height.
https://github.com/vim/vim/commit/d326ad6e93604222c64bae51f47521fe9114d1ea

janlazo committed:
vim-patch:8.0.1142: window toolbar menu gets a tear-off item

Problem:    Window toolbar menu gets a tear-off item.
Solution:   Recognize the window toolbar.
https://github.com/vim/vim/commit/378daf87d380b9f3c4f822786dfbfdcad9ca2db9

janlazo committed:
vim-patch:8.0.1138: click in window toolbar starts Visual mode

Problem:    Click in window toolbar starts Visual mode.
Solution:   Add the MOUSE_WINBAR flag.
https://github.com/vim/vim/commit/eb163d73b11c10b461a2839530173a33d7915a33

janlazo committed:
vim-patch:8.0.1292: quick clicks in the WinBar start Visual mode

Problem:    Quick clicks in the WinBar start Visual mode.
Solution:   Use a double click in the WinBar like a normal click.
https://github.com/vim/vim/commit/d327b0c68fa9b32877123f77560f4521a9eb39db

janlazo committed:
vim-patch:8.0.1334: splitting a window with a WinBar damages window layout

Problem:    Splitting a window with a WinBar damages window layout.
            (Lifepillar)
Solution:   Take the winbar into account when computing the new window
            position.  Add WINBAR_HEIGHT().
https://github.com/vim/vim/commit/3167c3e7010ac7e2fd7d3aa198dd2d1dcff8ecc8

janlazo committed:
vim-patch:8.0.1139: using window toolbar changes state

Problem:    Using window toolbar changes state.
Solution:   Always execute window toolbar actions in Normal mode.
https://github.com/vim/vim/commit/a21a6a9ade7bec3a07992d4d900d4ce82eeb8a29

teto commented:
Time to rebase :)

teto commented:
thanks the patches are applied correctly IIRC. I am on the fence on the w_winbar_height code. Will try to test vim to check what i is first.

janlazo committed:
vim-patch:8.1.1264: crash when closing window from WinBar click

Problem:    Crash when closing window from WinBar click. (Ben Jackson)
Solution:   Check that window pointer is still valid. (closes vim/vim#4337)
https://github.com/vim/vim/commit/d2fad67e3eb71f48d23f283ef8e7b7ddf4ee180f

janlazo commented:
WinBar feature is for Vim's `:TermDebug`, https://github.com/vim/vim/blob/master/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim. Most of it was ported to neovim but it's hard to tell because it was ported before merging the required toolbar patches in this PR. There are some Vim patches that were merged in Neovim's port of `TermDebug` but they're not documented in the commit history.

https://github.com/lifepillar/vim-colortemplate uses the toolbar feature so you can try that with nightly Vim to know how it works.

I'm removing the winbar code after all toolbar patches are committed in this PR or in master branch.

janlazo committed:
vim-patch:8.0.1375: window size wrong after maximizing with WinBar

Problem:    Window size wrong after maximizing with WinBar. (Lifepillar)
Solution:   Fix height computations. Redraw window when it is zero height but
            has a WinBar. (closes vim/vim#2356)
https://github.com/vim/vim/commit/415a6939a4e8d4e26b4af26c24eb75243d3a2756
